Implement UploadPyPi class

Duroktar opened the pull request:
- build method now builds dists according to a default or provided config.

- deploy method deploys to PyPI

- create helper functions for classes run in main entry point.

- tidy up entry point.

Duroktar committed:
Implement UploadPyPi class

- build method now builds dists according to a default or provided config.

- deploy method deploys to PyPI

- create helper functions for classes run in main entry point.

-tidy up entry point.

illume commented:
Wow. *Happy dance.* I have to go out now, but will have more of a look later tonight.

Cheers!

Duroktar commented:
Good stuff! Okay man, I'm resolving a couple of minor issues that popped up (of course :P) but we'll definitely talk later.

Cheers

Duroktar committed:
Add files via upload

Duroktar committed:
Add PyPi test site to twine options.

- Package can now be uploaded to the PyPi test site which
will make testing a lot easier.

- used on itself and works.

Test site link -
https://testpypi.python.org/pypi/pyrelease

illume commented:
I like how there are separate build commands.

Nice work on using the test pypi :)

What's the idea with the pybuilder config files? Is that a tool you use already? I haven't seen that before :) Checking it out. If we have config for the 'pyrelease' tool, it would be good to use the setup.cfg file, and use sections in it.

It looks like modern python uses https on pypi. So, I think we could drop the twine dependency? I checked at least my copy of python 2.7 and 3.6.
/usr/local/Cellar/python3/3.6.0/Frameworks/Python.framework/Versions/3.6/lib/python3.6/distutils/config.py.
/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/config.py

Duroktar commented:
Hey!

I've gotta get to bed soon but I want to respond because I was pretty excited to hear back from ya today.

The config file is a feature of trabBuild, the tool I made last week to build and distro packages. I merely threw it in there for consideration, as I understand your approach is a *no config* deal it can easily be omitted. The `import yaml` statement can be removed from the top of the file in that case.

Twine can certainly be dropped, that's no problem.

I'll modify and commit a new patch for this stuff tomorrow.

Any ideas where you'd like to go from here? I wouldn't mind starting work on a cli interface for gathering some basic user info at build time, what do you think?

Cheers and thanks,
  Scott


P.S. I'll pull pypackage from the test repo as well so you can have a chance to try it with your account.

illume commented:
Can you expand on what you mean by a cli interface? A good --help is really important. (maybe that can be written as the top doc string?).

Ok... random thoughts for where to go next below :)

I think user feedback is an important step. So, adding logging for everything. eg. 'Found author in .hgrc'. 'Creating temporary build folder at :%s:'. As well as negative messages like "Couldn't find a .gitrc file for author."

If there is a setup.py file, we should stop with a nice error message about how this tool doesn't work with setup.py files.

Nice error messages for setup upload would be good too. I noticed that if you tried to upload to a package that exists it gives you a '403 forbidden' error message. It took me some debugging to figure that out.

The error messages should all have a link to a url for documentation explaining things. Also the messages should contain enough information about things. Eg, we could print out the format for the .pypirc file so they can see it nicely.

Config in the file for \_\_author\_\_ \_\_license\_\_ etc would be good to support? Generating the LICENSE file. eg, if license is mit, download the mit one to include in the package.

Generating changes logs from git. Tagging releases in git. Getting version numbers from pypi.

I think if we get these basics down really well the tool would be quite useful for more people. Well, it can already release itself! That's pretty cool.


I won't commit anything on it for the next few days, because I'm working on a separate website project mostly.

illume commented:
Oh. Also, important for me is supporting `data/` folders, and supporting projects contained within packages `mypackage/`.

Duroktar commented:
All very good ideas. The cli would be the command line interface, so essentially the things that you mentioned above are all part of that. Since you'll be busy elsewhere for the next few days I'd be happy to start working on it. 

Cheers.

Duroktar commented:
Been trying to get in contact with you but can't find an e-mail anywhere. I have quite a bit of work done and would like to talk to you about it before I send another pull request..

- CLI implemented with help for each command
- Builds readme file from docs using slightly modified version of [doc2md](https://github.com/coldfix/doc2md/blob/master/doc2md.py)
- Minor tweaks to setup.py file creation
- Added very basic logging functionality, logging verbosity can be set from the CLI.
- Can set build path, or temporary file path for rendered project files.
- CLI option to deploy to PyPI test site instead of regular site.
- Can set version number from CLI now.
- Project is now organized into seperate modules ie:
 -> templates in their own spot
 -> User info classes are found under models.
 -> Utility functions can now be found under _utils.py
 -> I have also included a test folder with a basic script that can be built for testing purposes.

Most of these features need some tweaking done to them, but it all works. I have tested most of it and it's pretty slick. 

P.S. I haven't got around to "supporting data/ folders, and supporting projects contained within packages mypackage/" yet, and I haven't changed the twine command over to pip either, but these things can be done pretty easily. Also no license auto-inclusion yet ( I admit, I tunneled a bit when I was working and should have checked back here more, but hey :D ).

Cheers.

illume commented:
renesd at the google email service :)

illume commented:
ps. sounds pretty nice :)

E3V3A commented:
@Duroktar 
Did you pull your repo? I can't find your fork.

@illume 

...and so? What happened?
